ocaml/testsuite/tests/templates/basic/bad_param_impl.ml

```diff
@@ -0,0 +1,18 @@
+(* TEST
+
+readonly_files = "bad_param_impl.mli bad_param_impl.reference"
+
+* setup-ocamlc.byte-build-env
+** ocamlc.byte
+flags = "-as-parameter"
+module = "bad_param_impl.mli"
+*** ocamlc.byte
+flags = ""
+module = "bad_param_impl.ml"
+ocamlc_byte_exit_status = "2"
+compiler_output = "bad_param_impl.output"
+**** skip
+reason = "error currently broken"
```

Comment:
Why is this error currently broken? It seems like it should work

Reply:
Hmm ... I had thought this gets fixed later, but it looks like not. The issue is that the code in `typemod.ml` that's supposed to raise `Cannot_implement_parameter` can never actually raise: if it's declared as a parameter unit, the call to `Env.read_signature` fails. Basically, `persistent_env.ml` is currently too good at catching inconsistencies. It can't tell the difference between "this shouldn't be a parameter unit because it appeared in the source and it's not a `-parameter`" and "this shouldn't be a parameter because we're implementing it."

Basically, in order to report the error nicely to the user, either:

1. `Persistent_env` needs to do less error checking; or
2. `Persistent_env` needs to be given more context so that it can produce the correct error; or
3. pather than raising a fatal error itself, `Persistent_env` needs to return some error value that `Typemod` can interpret itself; or
4. in this particular case, I suppose `Persistent_env` could notice that the `.cmi` that shouldn't be a parameter happens to have the current module's name, and so it could assume that it should raise `Cannot_implement_parameter` instead of `Illegal_import_of_parameter`, but that feels a bit sketchy and I'm not sure it's always going to work out that nicely.

Reply:
I think you should take option 4. The code already looks up the current module name a few lines before the error is emitted.